### PR TITLE
V2 API

### PR DESCRIPTION
**Pull request by harry1453**
This is the first implementation of the V2 API, which currently contains the following calls:

```
    rpc GetBlock(GetBlockRequest) returns (Block) {}
    rpc GetAccount(GetAccountRequest) returns (Account) {}
    rpc GetAccounts(GetAccountsRequest) returns (Accounts) {}
    rpc GetMiningInfo(google.protobuf.Empty) returns (stream MiningInfo) {}
    rpc GetTransaction(GetTransactionRequest) returns (Transaction) {}
    rpc GetTransactionBytes(GetTransactionRequest) returns (TransactionBytes) {}
    rpc SubmitNonce(SubmitNonceRequest) returns (SubmitNonceResponse) {}
```

This means it is ready to use for mining (it contains every call that mining software / pool software could need).

The main advantage of the new API is speed, especially for miners. Instead of polling for new generation signatures, they can subscribe to be pushed to whenever a new block is forged (and a new mining round has begun). The other speed benefit is efficiency that is gained from using protocol buffers.

### CODE REVIEW

File: src/brs/grpc/handlers/SubmitNonceHandler.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    @Override
+    public BrsApi.SubmitNonceResponse handleRequest(BrsApi.SubmitNonceRequest request) throws Exception {
+        String secret = request.getSecretPhrase();
+        long nonce = request.getNonce();
+        long accountId = request.getAccount();
+        int submissionHeight = request.getBlockHeight();
+
+        if (submissionHeight != 0) {
+            if (submissionHeight != blockchain.getHeight() + 1) {
```
Comment (ohager) [inline src/brs/grpc/handlers/SubmitNonceHandler.java]: could join the ifs

File: src/brs/grpc/handlers/SubmitNonceHandler.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        if (Objects.equals(secret, "")) {
+            throw new ApiException("Missing Passphrase");
+        }
+
+        byte[] secretPublicKey = Crypto.getPublicKey(secret);
+        Account secretAccount = accountService.getAccount(secretPublicKey);
+        if(secretAccount != null) {
+            Account genAccount;
```
Comment (ohager) [inline src/brs/grpc/handlers/SubmitNonceHandler.java]: I tend to extract larger blocks in separate functions with crisp names to improve readability

File: src/brs/grpc/proto/BrsService.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class BrsService extends BrsApiServiceGrpc.BrsApiServiceImplBase {
+
+    private final Map<Class<? extends GrpcApiHandler<?,?>>, GrpcApiHandler<?,?>> handlers;
+
+    public BrsService(BlockchainProcessor blockchainProcessor, Blockchain blockchain, BlockService blockService, AccountService accountService, Generator generator, TransactionProcessor transactionProcessor) {
```
Comment (ohager) [inline src/brs/grpc/proto/BrsService.java]: nice

File: src/brs/grpc/proto/BrsService.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        GrpcApiHandler<?, ?> handler = handlers.get(handlerClass);
+        if (!handlerClass.isInstance(handler)) {
+            throw new HandlerNotFoundException();
+        }
+        return handlerClass.cast(handler);
+    }
+
+    @Override
+    public void getMiningInfo(Empty request, StreamObserver<BrsApi.MiningInfo> responseObserver) {
+        try {
```
Comment (ohager) [inline src/brs/grpc/proto/BrsService.java]: a lot of ctrl-c, ctrl-v --- yes, it's straightforward code, but boilerplate... maybe, would be an idea to consider Lamda functions to make it more fancy :yum:

File: src/brs/Attachment.java [hunk 5]
```diff
@@ -2018,6 +2284,15 @@ void putMyJSON(JsonObject attachment) {
...
 
+    @Override
+    public Any getProtobufMessage() {
+      return Any.pack(BrsApi.ATCreationAttachment.newBuilder()
+              .setVersion(getVersion())
+              .setName(name)
+              .setDescription(description)
+              .setCreationBytes(ByteString.copyFrom(creationBytes))
+              .build());
+    }
```
[the remark this answers is not in the document]
Comment (harry1453) [reply src/brs/Attachment.java]: I want to but this is outside the scope of this PR

File: src/brs/grpc/brsApi.proto [hunk 6]
```diff
@@ -0,0 +1,342 @@
+syntax = "proto3";
+import "google/protobuf/empty.proto";
+import "google/protobuf/any.proto";
+option java_package = "brs.grpc.proto";
+
+
+service BrsApiService {
+    rpc GetBlock(GetBlockRequest) returns (Block) {}
```
[the remark this answers is not in the document]
Comment (harry1453) [reply src/brs/grpc/brsApi.proto]: Nice!

File: src/brs/grpc/proto/BrsService.java (hunk 4 above)
Comment (harry1453) [reply src/brs/grpc/proto/BrsService.java]: I'll see what I can do 😄

File: Dockerfile.h2 [hunk 7]
```diff
@@ -1,13 +1,16 @@
 FROM maven:3-jdk-8-alpine as builder
+RUN apk update && apk add protobuf
+RUN apk add grpc-java --repository=http://dl-cdn.alpinelinux.org/alpine/edge/testing
 WORKDIR /app
 COPY . .
-RUN ./burst.sh compile
+RUN mvn -DskipTests=true -P headlessBuild package
```
Comment (blankey1337) [inline Dockerfile.h2]: skipTests ?

Comment (harry1453) [reply Dockerfile.h2]: Why waste time testing when building a docker image?

**Pull request merged**